vmTools: add in 9 RHEL-adjacent distros, upgrade Debian/Ubuntu

philiptaron opened the pull request:
This PR focuses on modernizing the `vmTools` infrastructure, improving reproducibility, and dramatically expanding the range of Linux distributions available for cross-compilation and virtualization testing.

## Core Infrastructure Changes

*   **Refactoring to use `callPackage`:** The core `vmTools` derivation was refactored to use `callPackage`. Now users can actually override specific packages used within the utility.
*   **Fixing Rotted Tests:** Old, non-functional tests referencing outdated distributions (Fedora 27, Ubuntu 18.04) were removed, and the remaining tests were updated to use the new `callPackage` structure. They're wired up in `tests.vmTools`.
*   **QEMU Utility Cleanup:** Internal QEMU utilities (`qemu-common`) were streamlined to accept `stdenv` directly, improving modularity, and these utilities are no longer exported by `vmTools`. (I could drop this commit -- it's sort of going sideways a bit.)
*   **FHS Compliance for Images:** Logic for creating disk images was updated to correctly handle modern merged directory structures (e.g., in Fedora 43), where `/bin`, `/sbin`, `/lib`, etc., are symlinked under `/usr`.
*   **Improved RPM Closure Handling:** The tool used to calculate required RPM packages (`rpm-closure.pl`) was updated to support modern compression formats like `.zst` (used by current Fedora releases) and now uses `URI::Escape` to correctly handle special characters in package URLs, ensuring functionality with distributions like Amazon Linux.

## Debian and Ubuntu Updates

The definitions for Debian and Ubuntu images were significantly updated to ensure stability and comprehensive package selection:

*   **Updated Debian Snapshots:** Debian 11 (Bullseye), Debian 12 (Bookworm), and Debian 13 (Trixie) snapshots were updated to the latest available timestamps (December 2025).
*   **Added Debian Backports:** The package lists for Debian 12 (Bookworm) and Debian 13 (Trixie) now include the `backports` repository, providing access to newer packages.
*   **Reproducible Ubuntu Sources:** Ubuntu 22.04 (Jammy) and 24.04 (Noble) images were switched from relying on potentially flaky `mirror://ubuntu` redirects to using fixed snapshot URLs from `https://snapshot.ubuntu.com`.
*   **Comprehensive Ubuntu Pockets:** Ubuntu images now include `updates` and `security` package pockets alongside the main packages, ensuring that images are built with essential maintenance and security fixes.

## Expanded RHEL-Family Distribution Support

Support for generating VM images for RHEL-family distributions was greatly expanded, focusing on modern, supported releases. This enables broader cross-platform compilation and testing.

A new internal `baseRHELFamilyPackages` list was introduced to centralize common build dependencies, reducing duplication across distro definitions.

New distributions added include:

| Distribution | Version(s) | Architecture | Notes |
| :--- | :--- | :--- | :--- |
| **Fedora** | 42, 43 | x86\_64 | Includes the latest releases. |
| **Rocky Linux** | 9, 10 | x86\_64 | Popular RHEL clone. |
| **Alma Linux** | 9, 10 | x86\_64 | Alternative RHEL clone. |
| **Oracle Linux** | 9 | x86\_64 | Official Oracle release. |
| **Amazon Linux** | 2023 | x86\_64 | Current version of Amazon's cloud distribution. |

Finally, **all** newly added distribution images are included in the `tests.vmTools` attrset.

I originally included CentOS Stream, but they only have an ephemeral store, so they rot extremely quickly.

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

philiptaron committed:
vmTools: use callPackage and fix rotted tests

philiptaron committed:
nixos: pass stdenv into qemu-common

We also don't export it from `vmTools` anymore.

philiptaron committed:
vmTools: update Debian 11 to snapshot 20251217T203845Z

philiptaron committed:
vmTools: update Debian 12 to snapshot 20251217T203845Z

philiptaron committed:
vmTools: update Debian 13 to snapshot 20251217T203845Z

philiptaron committed:
vmTools: update Ubuntu versions to use the snapshot.ubuntu.com server

philiptaron committed:
vmTools: add in updates and security pockets for all Ubuntu distros

philiptaron committed:
vmTools: add in backports packages for bookworm and trixie

philiptaron committed:
vmTools: add in Fedora 42 support

philiptaron committed:
vmTools: introduce Fedora 43

It uses unified sbin, which requires a slight tweak in the unifiedSystemDir setup

philiptaron committed:
vmTools: introduce CentOS Stream 9

philiptaron committed:
vmTools: introduce CentOS Stream 10

philiptaron committed:
vmTools: remove unused commonCentOSPackages

philiptaron committed:
vmTools: refactor in a baseRHELFamilyPackages

philiptaron committed:
vmTools: add Rocky Linux 9

philiptaron committed:
vmTools: add Rocky Linux 10

philiptaron committed:
vmTools: add Alma Linux 9 and 10

philiptaron committed:
vmTools: add Oracle Linux 9

philiptaron committed:
vmTools: add in Amazon Linux 2023

philiptaron committed:
vmTools: shore up the tests with each image

philiptaron commented:
That's super lame, Rocky. I'll take a look.

uninsane reviewed (changes requested):
could you update `doc/build-helpers/special/vm-tools.section.md` to reflect the changes? particularly the `#vm-tools-diskImageFuns-images` section.

maybe worth adding aliases like `ubuntuLatestx86_64` etc to the vmTools as a convenience, and using those in the docs to decrease the maintenance burden?

philiptaron committed:
doc: update vmTools documentation based on the newly added disk images

philiptaron commented on nix-in-debian.nix:
```diff
@@ -0,0 +1,110 @@
+let
```
Gotta delete this.

uninsane commented on pkgs/build-support/vm/default.nix:
```diff
@@ -985,25 +1014,287 @@ rec {
+      name = "fedora-42-x86_64";
+      fullName = "Fedora 42 (x86_64)";
+      packagesList = fetchurl {
+        url = "https://dl.fedoraproject.org/pub/fedora/linux/releases/42/Everything/x86_64/os/repodata/cd483b35df017d68b73a878a392bbf666a43d75db54c386e4720bc369eb5c3a3-primary.xml.zst";
```
https://dl.fedoraproject.org/pub/fedora/linux/releases/40/

looks like Fedora deletes files from `dl.fedoraproject.org` once the release goes EOL (so e.g. this URL probably has a lifespan < 1yr at this point) -- and moves them to a separate archive location.

interestingly, Fedora 41 is available both on `dl.fedoraproject.org` _and_ on the archive ([here](https://archives.fedoraproject.org/pub/archive/fedora/linux/updates/41)). i haven't compared those directories, but if the archives are a 1-to-1 copy of whatever we're grabbing, maybe we can paper over the instability by just listing both URLs here?

uninsane commented on pkgs/build-support/vm/default.nix:
```diff
@@ -985,25 +1014,287 @@ rec {
+      fullName = "Oracle Linux 9 (x86_64)";
+      packagesLists = [
+        (fetchurl {
+          url = "https://yum.oracle.com/repo/OracleLinux/OL9/baseos/latest/x86_64/repodata/edda5aa6dc1ad233cd5fe29838e535ef959cd291be343844e127e29c290f93ca-primary.xml.gz";
```
also short-lived. i believe you can use the dot-releases though, e.g. <https://yum.oracle.com/repo/OracleLinux/OL9/7/baseos/base/x86_64/repodata/>. Oracle seems to do a good job keeping those around, e.g. [6.1 from 2013](https://yum.oracle.com/repo/OracleLinux/OL6/1/base/x86_64/repodata/) is still up.

uninsane commented on pkgs/build-support/vm/default.nix:
```diff
@@ -1015,15 +1300,31 @@ rec {
         (fetchurl {
-          url = "mirror://ubuntu/dists/jammy/main/binary-amd64/Packages.xz";
-          sha256 = "sha256-N8tX8VVMv6ccWinun/7hipqMF4K7BWjgh0t/9M6PnBE=";
+          url = "https://snapshot.ubuntu.com/ubuntu/20251217T000000Z/dists/jammy/main/binary-amd64/Packages.xz";
```
i can believe that ubuntu 22.04 only ever got one snapshot and that's why every timestamp gives the same data. [snapshots.ubuntu.com](https://snapshot.ubuntu.com/) says it was introduced in 2023-05, and they seem to still be hosting [files for older releases like focal](https://snapshot.ubuntu.com/ubuntu/20230101T000000Z/dists/focal/main/binary-amd64/Packages.xz): these URLs should be fine.

philiptaron commented:
Thanks, this is a really useful review Colin. Round 3 incoming.

philiptaron committed:
vmTools: use AlmaLinux vault URLs for stable package lists

The repo.almalinux.org URLs are rolling and change with each minor
release, causing hash mismatches. Switch to vault.almalinux.org which
provides stable URLs for specific versions.

- alma9x86_64: use vault 9.6
- alma10x86_64: use vault 10.0

philiptaron committed:
vmTools: update Oracle Linux 9 package list hashes

Oracle provides versioned URLs for baseos but not appstream. Since we
can't mix versioned baseos with rolling appstream due to package version
dependencies (glibc mismatch), we keep using rolling URLs for both.

Add comment explaining this limitation.

philiptaron committed:
vmTools: add documentation for updating Amazon Linux 2023 GUID

Amazon Linux 2023 uses GUID-based URLs that don't allow directory
listing, making updates non-trivial. Add a comment explaining how to
find the current GUID by either running an AL2023 container or checking
the docker-library/repo-info repository.

philiptaron committed:
vmTools: drop CentOS Stream

As @uninsane points out, these are designed to rot rapidly.

philiptaron committed:
vmTools: update Rocky Linux 9.5 to 9.6

philiptaron committed:
vmTools: update Oracle Linux 9 package list hashes

philiptaron committed:
vmTools: update Debian snapshot timestamps to 20260105T082626Z

philiptaron committed:
vmTools: update Ubuntu snapshot timestamps to 20260101T000000Z
